front/components/agent_builder/capabilities/mcp/MCPServerViewsDialog.tsx

```diff
@@ -0,0 +1,694 @@
+  const onClickDataVisualization = () => {
+    if (!dataVisualization) {
+      return;
+    }
+    toggleToolSelection({ type: "DATA_VISUALIZATION" });
+  };
+
+  function onClickMCPServer(mcpServerView: MCPServerViewType) {
+    const tool = { type: "MCP", view: mcpServerView } satisfies SelectedTool;
+    const requirement = getMCPServerRequirements(mcpServerView);
+
+    if (!requirement.noRequirement) {
+      const action = getDefaultMCPAction(mcpServerView);
+      const isReasoning = requirement.requiresReasoningConfiguration;
+
+      if (action.type === "MCP" && isReasoning) {
+        if (reasoningModels.length === 0) {
+          sendNotification({
+            title: "No reasoning model available",
+            description:
+              "Please add a reasoning model to your workspace to be able to use this tool",
+            type: "error",
+          });
+          return;
+        }
+
+        const defaultReasoningModel =
+          reasoningModels.find(
+            (model) => model.modelId === DEFAULT_REASONING_MODEL_ID
+          ) ?? reasoningModels[0];
+
+        setConfigurationTool({
+          ...action,
+          configuration: {
+            ...action.configuration,
+            reasoningModel: {
+              modelId: defaultReasoningModel.modelId,
+              providerId: defaultReasoningModel.providerId,
+              temperature: null,
+              reasoningEffort: null,
+            },
+          },
+        });
+      } else {
+        setConfigurationTool(action);
+      }
+
+      setConfigurationMCPServerView(mcpServerView);
+      setCurrentPageId(CONFIGURATION_DIALOG_PAGE_IDS.CONFIGURATION);
+      return;
+    }
+
+    // No configuration required, add to selected tools
+    toggleToolSelection(tool);
+  }
+
+  const handleAddSelectedTools = useCallback(async () => {
+    // Validate any configured tools before adding
+    for (const tool of selectedToolsInDialog) {
+      if (tool.type === "MCP" && tool.configuredAction) {
+        const validation = validateMCPActionConfiguration(
+          tool.configuredAction,
+          tool.view
+        );
+
+        if (!validation.isValid) {
+          sendNotification({
+            title: "Configuration validation failed",
+            description: validation.errorMessage!,
+            type: "error",
+          });
+          return;
+        }
+      }
+    }
+
+    // All validations passed, add the tools
+    selectedToolsInDialog.forEach((tool) => {
+      if (tool.type === "DATA_VISUALIZATION") {
+        addTools({
+          id: uniqueId(),
+          type: "DATA_VISUALIZATION",
+          configuration: null,
+          name: DEFAULT_DATA_VISUALIZATION_NAME,
+          description: DEFAULT_DATA_VISUALIZATION_DESCRIPTION,
+          noConfigurationRequired: true,
+        });
+      } else if (tool.type === "MCP") {
+        const action = tool.configuredAction || getDefaultMCPAction(tool.view);
+        addTools(action);
+      }
+    });
```

Comment:
`addTools` seems to always be used in a loop. What about to avoid some re-rendering to use the `replace` instead of `append` coming from the Form?

Reply:
We can actually append all of them at once!